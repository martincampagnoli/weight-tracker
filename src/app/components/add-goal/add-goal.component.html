<form [formGroup]="formGroup">
  <mat-form-field class="pe-1 pt-1 w-100">
    <mat-label>kgs</mat-label>
    <input matInput formControlName="targetWeight" required />
  </mat-form-field>
</form>
<div class="pt-2">
  <button mat-button="tonal" (click)="reset()" class="me-1">
    Reset current goal
  </button>
  <button mat-button mat-raised-button (click)="setGoal()">Set new goal</button>
</div>
