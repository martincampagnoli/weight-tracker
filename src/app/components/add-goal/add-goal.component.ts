import {
  ChangeDetectionStrategy,
  Component,
  inject,
  input,
} from '@angular/core';
import {
  FormBuilder,
  FormsModule,
  ReactiveFormsModule,
  Validators,
} from '@angular/forms';
import * as reducers from 'src/app/store/default';
import { CommonModule } from '@angular/common';
import { MatButton } from '@angular/material/button';
import { MatInputModule, MatLabel } from '@angular/material/input';
import { MatDialogRef } from '@angular/material/dialog';
import { MatFormFieldModule } from '@angular/material/form-field';
import { Store } from '@ngrx/store';
import { resetGoal, setGoal } from 'src/app/store/default';
import { SnackBarUtil } from 'src/app/utils/snackbar.util';

@Component({
  selector: 'app-add-goal',
  templateUrl: './add-goal.component.html',
  styleUrls: ['./add-goal.component.scss'],
  standalone: true,
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
    MatFormFieldModule,
    MatInputModule,
    MatLabel,
    MatButton,
  ],
})
export class AddGoalComponent {
  private formBuilder = inject(FormBuilder);
  private snackBarUtil: SnackBarUtil = inject(SnackBarUtil);
  private store: Store<reducers.AppState> = inject(Store);
  readonly dialogRef = input.required<MatDialogRef<AddGoalComponent>>();

  protected formGroup = this.formBuilder.group({
    targetWeight: ['', [Validators.required]],
  });

  protected setGoal(): void {
    const targetWeight = Number(this.formGroup.value.targetWeight);

    if (!this.formGroup.valid) {
      return;
    }

    this.store.dispatch(
      setGoal({
        payload: {
          targetWeight,
        },
      })
    );
    this.snackBarUtil.show('Goal set successfully');
    this.dialogRef().close();
  }

  protected reset(): void {
    this.store.dispatch(resetGoal());
    this.dialogRef().close();
  }
}
