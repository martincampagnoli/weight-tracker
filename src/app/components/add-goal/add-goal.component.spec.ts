import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { provideMockStore } from '@ngrx/store/testing';
import { MatDialogRef } from '@angular/material/dialog';
import { MatSnackBar } from '@angular/material/snack-bar';
import { provideNativeDateAdapter } from '@angular/material/core';
import { signal } from '@angular/core';
import { AddGoalComponent } from './add-goal.component';

describe('AddGoalComponent', () => {
  let component: AddGoalComponent;
  let fixture: ComponentFixture<AddGoalComponent>;
  let mockDialogRef: jest.Mocked<MatDialogRef<AddGoalComponent>>;
  let mockSnackBar: jest.Mocked<MatSnackBar>;

  const initialState = {
    data: [],
  };

  beforeEach(async () => {
    mockDialogRef = {
      close: jest.fn(),
      afterClosed: jest.fn(),
      beforeClosed: jest.fn(),
    } as any;

    mockSnackBar = {
      open: jest.fn(),
      dismiss: jest.fn(),
    } as any;

    await TestBed.configureTestingModule({
      imports: [AddGoalComponent, NoopAnimationsModule],
      providers: [
        provideMockStore({ initialState }),
        provideNativeDateAdapter(),
        { provide: MatSnackBar, useValue: mockSnackBar },
        { provide: 'dialogRef', useValue: mockDialogRef },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(AddGoalComponent);
    component = fixture.componentInstance;

    fixture.componentRef.setInput('dialogRef', signal(mockDialogRef));
    fixture.detectChanges();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Component Creation', () => {
    it('should create', () => {
      expect(component).toBeTruthy();
    });

    it('should match snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
});
