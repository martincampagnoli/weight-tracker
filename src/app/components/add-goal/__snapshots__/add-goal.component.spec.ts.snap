// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddGoalComponent Component Creation should match snapshot 1`] = `
<app-add-goal
  dialogRef={[Function Function]}
  formBuilder={[Function _FormBuilder]}
  formGroup={[Function FormGroup]}
  snackBarUtil={[Function SnackBarUtil]}
  store={[Function _MockStore]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <mat-form-field
      class="mat-mdc-form-field pe-1 pt-1 w-100 mat-mdc-form-field-type-mat-input mat-form-field-appearance-fill mat-form-field-hide-placeholder mat-primary ng-untouched ng-pristine ng-invalid"
    >
      <div
        class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled"
      >
        <div
          class="mat-mdc-form-field-focus-overlay"
        />
        <div
          class="mat-mdc-form-field-flex"
        >
          <div
            class="mat-mdc-form-field-infix"
          >
            <label
              class="mdc-floating-label mat-mdc-floating-label"
              for="mat-input-a1"
              id="mat-mdc-form-field-label-a1"
              matformfieldfloatinglabel=""
            >
              <mat-label>
                kgs
              </mat-label>
              <span
                aria-hidden="true"
                class="mat-mdc-form-field-required-marker mdc-floating-label--required"
              />
            </label>
            <input
              aria-required="true"
              class="mat-mdc-input-element ng-untouched ng-pristine ng-invalid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
              formcontrolname="targetWeight"
              id="mat-input-a1"
              matinput=""
              required=""
            />
          </div>
        </div>
        <div
          class="mdc-line-ripple mdc-line-ripple--deactivating"
          matformfieldlineripple=""
        />
      </div>
      <div
        aria-atomic="true"
        aria-live="polite"
        class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
      >
        <div
          class="mat-mdc-form-field-hint-wrapper"
        >
          <div
            class="mat-mdc-form-field-hint-spacer"
          />
        </div>
      </div>
    </mat-form-field>
  </form><div
    class="pt-2"
  >
    <button
      class="mdc-button mat-mdc-button-base me-1 mat-mdc-button mat-unthemed _mat-animation-noopable"
      mat-button="tonal"
      mat-ripple-loader-class-name="mat-mdc-button-ripple"
      mat-ripple-loader-uninitialized=""
    >
      <span
        class="mat-mdc-button-persistent-ripple mdc-button__ripple"
      />
      <span
        class="mdc-button__label"
      >
         Reset current goal 
      </span>
      <span
        class="mat-focus-indicator"
      />
      <span
        class="mat-mdc-button-touch-target"
      />
    </button>
    <button
      class="mdc-button mat-mdc-button-base mdc-button--raised mat-mdc-raised-button mat-unthemed _mat-animation-noopable"
      mat-button=""
      mat-raised-button=""
      mat-ripple-loader-class-name="mat-mdc-button-ripple"
      mat-ripple-loader-uninitialized=""
    >
      <span
        class="mat-mdc-button-persistent-ripple mdc-button__ripple"
      />
      <span
        class="mdc-button__label"
      >
        Set new goal
      </span>
      <span
        class="mat-focus-indicator"
      />
      <span
        class="mat-mdc-button-touch-target"
      />
    </button>
  </div>
</app-add-goal>
`;
